File: src-tauri/src/connectors/registry.rs
use crate::{application::ports::DatabaseConnectorFactory, domain::DbKind};

use super::OracleConnector;

/// Resolves a database kind to its port-compatible connector factory.
///
/// The registry owns concrete adapters so application services only ever see
/// the database ports, not a driver's connection or value types.
pub struct ConnectorRegistry {
    oracle: Box<dyn DatabaseConnectorFactory>,
}

impl ConnectorRegistry {
    pub fn with_oracle(oracle: OracleConnector) -> Self {
        Self {
            oracle: Box::new(oracle),
        }
    }

    pub fn for_kind(&self, kind: DbKind) -> Option<&dyn DatabaseConnectorFactory> {
        match kind {
            DbKind::Oracle => Some(self.oracle.as_ref()),
        }
    }
}

File: electron/connectors/registry.ts
import type { DbKind } from "../domain/models";
import { ConnectorError, type DatabaseConnectorFactory } from "./databaseConnector";
import { OracleConnector } from "./oracleConnector";

export class ConnectorRegistry {
  private readonly connectors: ReadonlyMap<DbKind, DatabaseConnectorFactory>;

  constructor(connectors: readonly DatabaseConnectorFactory[] = [new OracleConnector()]) {
    this.connectors = new Map(connectors.map((connector) => [connector.kind, connector]));
  }

  forKind(kind: DbKind): DatabaseConnectorFactory {
    const connector = this.connectors.get(kind);
    if (connector === undefined) {
      throw new ConnectorError("CONNECTOR_NOT_FOUND", `connector not found for ${kind}`);
    }
    return connector;
  }
}

export type { DatabaseConnectorFactory, DatabaseSession } from "./databaseConnector";
export { ConnectorError } from "./databaseConnector";
